src/infer_request.cc

```diff
@@ -1006,12 +1006,50 @@ InferenceRequest::Normalize()
     // If no input is marked as optional, then use exact match error message
     // for consistency / backward compatibility
     if ((size_t)model_config.input_size() == model_raw_->RequiredInputCount()) {
+      // Model config is master
+      // Original inputs is smaller
+      std::string missing_input_string = "[";
+      std::string original_input_string = "[";
+
+      int missing_input_cnt = 0;
+      for (size_t i = 0; i < (size_t)model_config.input_size(); ++i) {
+        const inference::ModelInput& input = model_config.input(i);
+        if (original_inputs_.find(input.name()) == original_inputs_.end()) {
+          missing_input_string =
+              missing_input_string + "'" + input.name() + "'" + ",";
+          missing_input_cnt++;
+        }
+      }
+      // Removes the extra ","
+      // Can be handled in a a more elegant way by creating a vector first of
+      // the inputs and then the string this way saves memory and overhead of
+      // creating a vector.
+      missing_input_string.pop_back();
+      missing_input_string = missing_input_string + "]";
+
+      for (const auto& pair : original_inputs_) {
+        original_input_string =
+            original_input_string + "'" + pair.first + "'" + ",";
+      }
+      // Removes the extra ","
+      // Can be handled in a a more elegant way by creating a vector first of
+      // the inputs and then the string this way saves memory and overhead of
+      // creating a vector.
+      original_input_string.pop_back();
+      original_input_string = original_input_string + "]";
+      if (missing_input_cnt == 0) {
+        missing_input_string = "[]";
+      }
+      if (original_inputs_.size() == 0) {
+        original_input_string = "[]";
+      }
```

Comment:
We'll want to test a few edge cases here in some new tests:
1. some required + some optional inputs
    - say a model has 2 required inputs and 2 optional inputs, but 1 required input is missing. Will the error message say something like `"Got inputs [required0], but missing [required1, optional0, optional1]"` or will it say `"Got inputs [required0], but missing [required1]"`. We may need some extra logic to distinguish between required and optional in these messages.

2. can `missing_count == 0` case happen? wouldn't this imply we had all the inputs?
3. should test `original_inputs_.size() == 0` just to automate validating we get the right "[]" empty string you added

Reply:
I can't find a good place to append these tests to, so I think we can just add a new test. Maybe named something like `L0_input_validation`. You can probably use [L0_nan_inf](https://github.com/triton-inference-server/server/tree/main/qa/L0_nan_inf) as a relatively simple reference for how to write a new test.

Basically you need 3 things:
1. `models/` - some example model repository containing simple models used to reproduce your error messages / test cases
2. `input_validation_test.py` - defines the unit tests for various cases, i.e. setting up clients, sending requests to different models, and validating the error messages returned
3. `test.sh` - starts tritonserver using your test model repository, then runs your python unit tests

We can chat more offline about the details of adding/writing tests if you have questions

Reply:
Taken care in the server repo PR
https://github.com/triton-inference-server/server/pull/6933

Reply:
Also added extra logic to handle optional inputs